x/ante/restricted_tx_test.go

```diff
@@ -0,0 +1,97 @@
+	"golang.org/x/exp/maps"
+
+	"github.com/axelarnetwork/axelar-core/testutils/rand"
+	"github.com/axelarnetwork/axelar-core/x/ante"
+	"github.com/axelarnetwork/axelar-core/x/ante/types/mock"
+	axelarnet "github.com/axelarnetwork/axelar-core/x/axelarnet/types"
+	evm "github.com/axelarnetwork/axelar-core/x/evm/types"
+	"github.com/axelarnetwork/axelar-core/x/permission/exported"
+	tss "github.com/axelarnetwork/axelar-core/x/tss/types"
+	"github.com/axelarnetwork/utils/slices"
+	. "github.com/axelarnetwork/utils/test"
+)
+
+func TestRestrictedTx(t *testing.T) {
+	var (
+		handler    ante.RestrictedTx
+		permission *mock.PermissionMock
+		tx         *mock.TxMock
+	)
+
+	signerAnyRole := func(t *testing.T) {
+		permission.GetRoleFunc = func(sdk.Context, sdk.AccAddress) exported.Role {
+			return exported.Role(rand.Of[int32](maps.Keys(exported.Role_name)...))
+		}
+	}
+
+	signerIsNot := func(role exported.Role) func(t *testing.T) {
+		return func(t *testing.T) {
+			permission.GetRoleFunc = func(sdk.Context, sdk.AccAddress) exported.Role {
+				filtered := slices.Filter(maps.Keys(exported.Role_name), func(k int32) bool { return k != int32(role) })
+				return exported.Role(rand.Of[int32](filtered...))
+			}
+		}
+	}
+
+	letTxThrough := func(t *testing.T) {
+		_, err := handler.AnteHandle(sdk.Context{}, tx, false,
+			func(sdk.Context, sdk.Tx, bool) (sdk.Context, error) { return sdk.Context{}, nil })
+		assert.NoError(t, err)
+	}
+
+	stopTx := func(t *testing.T) {
+		_, err := handler.AnteHandle(sdk.Context{}, tx, false,
+			func(sdk.Context, sdk.Tx, bool) (sdk.Context, error) { return sdk.Context{}, nil })
+		assert.Error(t, err)
+	}
+
+	txWithMsg := func(msg descriptor.Message) *mock.TxMock {
+		return &mock.TxMock{
+			GetMsgsFunc: func() []sdk.Msg {
+				return slices.Expand(func(_ int) sdk.Msg {
+					return &mock.MsgMock{
+						GetSignersFunc: func() []sdk.AccAddress {
+							return slices.Expand(func(_ int) sdk.AccAddress { return rand.AccAddr() }, int(rand.I64Between(1, 5)))
+						},
+						DescriptorFunc: msg.Descriptor,
+					}
+				}, int(rand.I64Between(1, 20)))
+			},
+		}
+	}
+
+	msgRoleIsUnrestricted := func(t *testing.T) { tx = txWithMsg(&evm.LinkRequest{}) }
+	msgRoleIsUnspecified := func(t *testing.T) { tx = txWithMsg(&banktypes.MsgSend{}) }
+	msgRoleIsChainManagement := func(t *testing.T) { tx = txWithMsg(&tss.RotateKeyRequest{}) }
+	msgRoleIsAccessControl := func(t *testing.T) { tx = txWithMsg(&axelarnet.RegisterFeeCollectorRequest{}) }
+
+	Given("a restricted tx ante handler", func(t *testing.T) {
+		permission = &mock.PermissionMock{}
+		handler = ante.NewRestrictedTx(permission)
+	}).Branch(
+		When("msg role is unrestricted", msgRoleIsUnrestricted).
+			And().When("signer has any role", signerAnyRole).
+			Then("let the msg through", letTxThrough),
+
+		When("msg role is unspecified", msgRoleIsUnspecified).
+			And().When("signer has any role", signerAnyRole).
+			Then("let the msg through", letTxThrough),
```

Comment:
what is the actual difference between an unspecified and an unrestricted node? Seems like we treat both types the same

Reply:
unspecified: the option is not set on the message (important for cosmos sdk and ibc messages)
unrestricted: explicitly set option to let anyone execute it